File: app/src/main/java/ru/samtakoy/core/business/CardsRepository.java
package ru.samtakoy.core.business;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Observable;
import ru.samtakoy.core.model.Card;
import ru.samtakoy.core.model.QPack;
import ru.samtakoy.core.model.Theme;

public interface CardsRepository {

    Completable clearDb();

    Card getCard(Long cardId);
    void saveCard(Card card);

    QPack getQPack(Long qPackId);
    void deletePack(Long qPackId);
    void saveQPack(QPack qPack);
    boolean hasAnyQPack();
    List<Card> getQPackCards(QPack qPack);
    int getQPackCardCount(Long qPackId);
    List<Card> getQPackCardsWithTags(QPack qPack);

    List<Theme> getChildThemes(Long themeId);
    List<QPack> getChildQPacks(Long themeId);

    Observable<QPack> getAllQPacks();

    Theme getTheme(Long themeId);
    Long addNewTheme(Long parentThemeId, String title);
    boolean deleteTheme(Long themeId);


    //Theme getParentTheme(Long themeId);

}

File: app/src/main/java/ru/samtakoy/core/business/CardsRepository.kt
package ru.samtakoy.core.business

import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import ru.samtakoy.core.database.room.entities.CardEntity
import ru.samtakoy.core.database.room.entities.other.CardIds
import ru.samtakoy.core.database.room.entities.other.CardWithTags

interface CardsRepository {

    fun clearDb(): Completable
    fun updateCard(card: CardEntity)
    fun deleteCard(cardId: Long)
    fun deleteQPackCards(qPackId: Long): Completable

    fun addCard(card: CardEntity): Long
    fun getCard(cardId: Long): CardEntity?
    fun getCardRx(cardId: Long): Flowable<CardEntity>
    fun getCardIds(cardId: Long): CardIds?


    fun getQPackCards(qPackId: Long): Flowable<List<CardEntity>>
    fun getCardsIdsFromQPack(qPackId: Long): Single<List<Long>>

    fun getQPackCardsWithTagsRx(qPackId: Long): Single<List<CardWithTags>>
    fun getQPackCardCount(qPackId: Long): Int


}